server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/DatafileTransactionLog.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.tserver.tablet;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import org.apache.accumulo.core.conf.AccumuloConfiguration;
+import org.apache.accumulo.core.conf.Property;
+import org.apache.accumulo.core.dataImpl.KeyExtent;
+import org.apache.accumulo.core.metadata.StoredTabletFile;
+import org.apache.accumulo.server.conf.TableConfiguration;
+
+public class DatafileTransactionLog {
+  private final KeyExtent extent;
+  private final Set<StoredTabletFile> initialFiles = new HashSet<>();
+  private final List<DatafileTransaction> tabletLogs =
+      Collections.synchronizedList(new LinkedList<>());
+  private final AccumuloConfiguration.Deriver<MaxLogSize> maxSize;
+  private volatile long initialTs = System.currentTimeMillis();
+
+  public DatafileTransactionLog(KeyExtent extent, Set<StoredTabletFile> initialFiles,
```

Comment:
This class is another source of locking on tablet operations (in addition to whatever the Tablet/DatafileManager already do). It'd be nice to be able to turn this off once we feel like we are no longer seeing the memory divergences. Right now, even if I turned the log size down to 0, then every call to compact, flush, or import are going to contend on the `tabletLogs` lock in this class. 

If we set the log length to 0, maybe we could bypass this class altogether? Or maybe just short circuit the methods.

Reply:
That is a valid point.  I will see if I can minimize the synchronization points.

Reply:
I minimized the synchronization in this class by using a separate immutable transaction log state class.